docs/modules/ROOT/pages/adr/0033-event-based-billing-oddo.adoc

```diff
@@ -0,0 +1,145 @@
+* Greater control over services when billing fails.
+* Billing logic is treated as a first-class part of provisioning/deprovisioning.
+
+*Cons:*
+
+* Persistence integration becomes more complex.
+* No clear separation between billing and service logic.
+* Unclear where billing actually occurs, which can reduce maintainability.
+
+== Persistence Options
+
+Based on past experience, we anticipate the need to resend older data to Odoo due to potential issues either in AppCat or Odoo.
+A lightweight, reliable mechanism to store and replay billing data is essential.
+
+Required capabilities:
+
+* Filtering and querying historical events.
+* Manual replay.
+* Partial delivery of historical events.
+
+=== SQLite
+
+SQLite is a simple, embedded SQL database engine suitable for local persistence needs.
+
+*Pros:*
+
+* Minimal setup; no external infrastructure required.
+* Fast for local and sequential read/write operations.
+* Full SQL support.
+* ACID-compliant (supports WAL mode).
+* Self-contained `.db` file that's easy to handle and back up.
+* Supports pagination and filtering by retry state or timestamp.
+
+*Cons:*
+
+* Not suitable for concurrent writes across multiple pods.
+* Requires manual effort for backups, failover, and compaction.
+* Not distributed or highly available.
+* Lacks integration with Kubernetes tools like `kubectl`.
+* Not inherently event-driven.
+
+=== Custom Kubernetes CRDs
+
+Custom Resource Definitions (CRDs) can be used to model billing events as native Kubernetes objects.
+
+*Pros:*
+
+* Native integration with Kubernetes and observable via `kubectl`.
+* Supports event-driven architectures through controllers.
+* State tracking via `.status` fields.
+* Reusable by other tools/controllers within the cluster.
+* Scales horizontally (no single-writer limitation).
+
+*Cons:*
+
+* Excessive CR volume may cause etcd bloat, impacting cluster performance.
+* Increased API server traffic.
+* Requires boilerplate for CRD definitions and status handling.
+* No native support for complex queries (unlike SQL).
+* Manual schema migration is necessary.
+* No built-in audit trail beyond resource versioning.
+
+== Recommendation
+
+=== Use Controller + Custom Kubernetes CRD
+
+The recommendation is to extend the existing AppCat controller to manage event-based billing and using Kubernetes CRDs as the persistence mechanism.
+
+*Justification:*
+
+* A controller is the natural place for billing, as it sits adjacent to service lifecycle management without coupling to it.
+* CRDs integrate well into our Kubernetes-native toolset and align with GitOps principles.
+* Data inspection and interaction via `kubectl` is simple and consistent with existing workflows.
+* While CRs are harder to query than SQL databases, we can mitigate this by providing predefined `kubectl` query templates for common tasks.
+* Kubernetes retry mechanisms can be leveraged for automatic re-delivery of failed events.
```

Comment:
We need to be careful here. Events are only present on a cluster for up to 1 hour. Therefore, we can't re-deliver anything older than 1 hour. This is especially true if the CR is deleted as we can't get any information from there either (eg. status field)

Reply:
We should also definitely set finalizers on those objects. So we can be sure they don't get deleted by accident.

Reply:
yes we need

Reply:
Added a finaliser in my CR example